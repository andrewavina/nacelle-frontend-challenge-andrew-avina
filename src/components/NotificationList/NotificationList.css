.notification-container {
  width: 400px;
  margin: 40px auto;
  font-family: Arial, sans-serif;
}

.add-button {
  display: block;
  margin: 0 auto 20px;
  padding: 10px 20px;
  background-color: #3f51b5;
  color: #ffffff;
  border: none;
  border-radius: 4px;
  font-size: 16px;
  cursor: pointer;
}

.add-button:hover {
  background-color: #303f9f;
}

.notification-list {
  list-style-type: none;
  padding: 0;
}

.notification-item {
  background-color: #f5f5f5;
  padding: 15px 20px;
  margin-bottom: 10px;
  border-radius: 4px;
  cursor: pointer;
  transition: background-color 0.2s ease;
}

.notification-item:hover {
  background-color: #e0e0e0;
}

.visually-hidden {
  position: absolute;
  width: 1px;
  height: 1px;
  margin: -1px;
  padding: 0;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  border: 0;
}
