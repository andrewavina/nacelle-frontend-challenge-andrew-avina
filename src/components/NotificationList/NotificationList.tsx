import React from "react";
import "./NotificationList.css";
import { useSelector, useDispatch } from "react-redux";
import { RootState } from "../../store";
import {
  addNotification,
  removeNotification,
} from "../../features/notifications/notificationsSlice";

const NotificationList: React.FC = () => {
  const notifications = useSelector(
    (state: RootState) => state.notifications.notifications,
  );
  const dispatch = useDispatch();

  const handleAddNotification = () => {
    const newNotification: AppNotification = {
      id: Date.now().toString(),
      message: `This is notification ${notifications.length + 1}`,
    };
    dispatch(addNotification(newNotification));
  };

  const handleRemoveNotification = (id: string) => {
    dispatch(removeNotification(id));
  };

  return (
    <section
      className="notification-container"
      aria-labelledby="notification-heading"
    >
      <h2 id="notification-heading" className="visually-hidden">
        Notifications
      </h2>
      <button className="add-button" onClick={handleAddNotification}>
        Add Notification
      </button>
      {notifications.length === 0 ? (
        <p className="empty-notification">No notifications</p>
      ) : (
        <ul className="notification-list" role="list">
          {notifications.map((notification) => (
            <li
              key={notification.id}
              className="notification-item"
              role="listitem"
              tabIndex={0}
              onClick={() => handleRemoveNotification(notification.id)}
              onKeyPress={(e) => {
                if (e.key === "Enter") {
                  handleRemoveNotification(notification.id);
                }
              }}
            >
              {notification.message}
            </li>
          ))}
        </ul>
      )}
    </section>
  );
};

export default NotificationList;
