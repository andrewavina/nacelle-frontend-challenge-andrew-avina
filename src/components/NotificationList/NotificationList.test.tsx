import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import NotificationList from "./NotificationList";
import {
  addNotification,
  removeNotification,
} from "../../features/notifications/notificationsSlice";

const mockStore = configureStore([]);
const initialState = {
  notifications: {
    notifications: [],
  },
};

describe("NotificationList component unit tests:", () => {
  let store: ReturnType<typeof mockStore>;

  beforeEach(() => {
    store = mockStore(initialState);
    store.dispatch = jest.fn();
  });

  test("renders NotificationList component", () => {
    render(
      <Provider store={store}>
        <NotificationList />
      </Provider>,
    );

    expect(screen.getByText("Add Notification")).toBeInTheDocument();
    expect(screen.getByText("No notifications")).toBeInTheDocument();
  });

  test("adds a notification", () => {
    render(
      <Provider store={store}>
        <NotificationList />
      </Provider>,
    );

    fireEvent.click(screen.getByText("Add Notification"));

    expect(store.dispatch).toHaveBeenCalledWith(
      addNotification(expect.any(Object)),
    );
  });

  test("removes a notification", () => {
    const stateWithNotifications = {
      notifications: {
        notifications: [{ id: "1", message: "Test Notification 1" }],
      },
    };
    store = mockStore(stateWithNotifications);

    render(
      <Provider store={store}>
        <NotificationList />
      </Provider>,
    );

    fireEvent.click(screen.getByText("Test Notification 1"));

    const actions = store.getActions();

    expect(actions).toContainEqual(removeNotification("1"));
  });

  test("renders notifications from the store", () => {
    const stateWithNotifications = {
      notifications: {
        notifications: [
          { id: "1", message: "Test Notification 1" },
          { id: "2", message: "Test Notification 2" },
        ],
      },
    };
    store = mockStore(stateWithNotifications);

    render(
      <Provider store={store}>
        <NotificationList />
      </Provider>,
    );

    expect(screen.getByText("Test Notification 1")).toBeInTheDocument();
    expect(screen.getByText("Test Notification 2")).toBeInTheDocument();
  });
});

describe("NotificationList component snapshot tests", () => {
  test("matches snapshot with no notifications", () => {
    const store = mockStore({ notifications: { notifications: [] } });
    const { asFragment } = render(
      <Provider store={store}>
        <NotificationList />
      </Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("matches snapshot with one notification", () => {
    const store = mockStore({
      notifications: {
        notifications: [{ id: "1", message: "Test Notification" }],
      },
    });
    const { asFragment } = render(
      <Provider store={store}>
        <NotificationList />
      </Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("matches snapshot with multiple notifications", () => {
    const store = mockStore({
      notifications: {
        notifications: [
          { id: "1", message: "Notification 1" },
          { id: "2", message: "Notification 2" },
        ],
      },
    });
    const { asFragment } = render(
      <Provider store={store}>
        <NotificationList />
      </Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
