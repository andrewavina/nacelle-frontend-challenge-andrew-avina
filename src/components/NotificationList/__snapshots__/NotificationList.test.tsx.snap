// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationList component snapshot tests matches snapshot with multiple notifications 1`] = `
<DocumentFragment>
  <section
    aria-labelledby="notification-heading"
    class="notification-container"
  >
    <h2
      class="visually-hidden"
      id="notification-heading"
    >
      Notifications
    </h2>
    <button
      class="add-button"
    >
      Add Notification
    </button>
    <ul
      class="notification-list"
      role="list"
    >
      <li
        class="notification-item"
        role="listitem"
        tabindex="0"
      >
        Notification 1
      </li>
      <li
        class="notification-item"
        role="listitem"
        tabindex="0"
      >
        Notification 2
      </li>
    </ul>
  </section>
</DocumentFragment>
`;

exports[`NotificationList component snapshot tests matches snapshot with no notifications 1`] = `
<DocumentFragment>
  <section
    aria-labelledby="notification-heading"
    class="notification-container"
  >
    <h2
      class="visually-hidden"
      id="notification-heading"
    >
      Notifications
    </h2>
    <button
      class="add-button"
    >
      Add Notification
    </button>
    <p
      class="empty-notification"
    >
      No notifications
    </p>
  </section>
</DocumentFragment>
`;

exports[`NotificationList component snapshot tests matches snapshot with one notification 1`] = `
<DocumentFragment>
  <section
    aria-labelledby="notification-heading"
    class="notification-container"
  >
    <h2
      class="visually-hidden"
      id="notification-heading"
    >
      Notifications
    </h2>
    <button
      class="add-button"
    >
      Add Notification
    </button>
    <ul
      class="notification-list"
      role="list"
    >
      <li
        class="notification-item"
        role="listitem"
        tabindex="0"
      >
        Test Notification
      </li>
    </ul>
  </section>
</DocumentFragment>
`;
